File: src/deuce/DottedPair.java
package deuce;

public class DottedPair {
    public Object car;
    public Object cdr;

    public DottedPair(Object car, Object cdr) {
        this.car = car;
        this.cdr = cdr;
    }

    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DottedPair that = (DottedPair) o;

        if (car != null ? !car.equals(that.car) : that.car != null) return false;
        if (cdr != null ? !cdr.equals(that.cdr) : that.cdr != null) return false;

        return true;
    }

    public int hashCode() {
        int result = car != null ? car.hashCode() : 0;
        result = 31 * result + (cdr != null ? cdr.hashCode() : 0);
        return result;
    }
}

File: src/deuce/dotted_pair.clj
(ns deuce.dotted-pair)

(defprotocol IDottedPair
  (car [this])
  (set-car! [this val])
  (cdr [this])
  (set-cdr! [this val]))

(deftype DottedPair [^:volatile-mutable fst ^:volatile-mutable snd]
  IDottedPair
  (car [this] fst)
  (set-car! [this val] (set! fst val))
  (cdr [this] snd)
  (set-cdr! [this val] (set! snd val))

  Object
  (equals [this that]
    (or (identical? this that)
        (and (instance? DottedPair that)
             (= fst (car that))
             (= snd (cdr that)))))
  (hashCode [_]
    (+ (* 31 (if (nil? fst) 0 (.hashCode fst)))
       (if (nil? snd) 0 (.hashCode snd)))))
